api/agent/protocol/cloudevent.go

```diff
@@ -0,0 +1,207 @@
+	in.Extensions["protocol"] = CallRequestHTTP{
+		Type:       ci.ProtocolType(),
+		Method:     ci.Method(),
+		RequestURL: ci.RequestURL(),
+		Headers:    ci.Headers(),
+	}
+	in.Extensions["deadline"] = ci.Deadline().String()
+
+	fmt.Printf("SENDING IN: %+v\n", in)
+	return json.NewEncoder(h.in).Encode(in)
+}
+
+func (h *CloudEventProtocol) Dispatch(ctx context.Context, ci CallInfo, w io.Writer) error {
+	ctx, span := trace.StartSpan(ctx, "dispatch_json")
+	defer span.End()
+
+	_, span = trace.StartSpan(ctx, "dispatch_json_write_request")
+	err := h.writeJSONToContainer(ci)
+	span.End()
+	if err != nil {
+		return err
+	}
+
+	_, span = trace.StartSpan(ctx, "dispatch_json_read_response")
+	var jout cloudEventOut
+	decoder := json.NewDecoder(h.out)
+	err = decoder.Decode(&jout)
+	span.End()
+	if err != nil {
+		return models.NewAPIError(http.StatusBadGateway, fmt.Errorf("invalid json response from function err: %v", err))
+	}
+
+	_, span = trace.StartSpan(ctx, "dispatch_json_write_response")
+	defer span.End()
+
+	rw, ok := w.(http.ResponseWriter)
+	if !ok {
+		// logs can just copy the full thing in there, headers and all.
+		err := json.NewEncoder(w).Encode(jout)
+		return isExcessData(err, decoder)
+	}
+
+	// this has to be done for pulling out:
+	// - status code
+	// - body
+	// - headers
+	pp := jout.Extensions["protocol"]
+	var p map[string]interface{}
+	if pp != nil {
+		p = pp.(map[string]interface{})
+		hh := p["headers"]
+		if hh != nil {
+			h, ok := hh.(map[string]interface{})
+			if !ok {
+				return fmt.Errorf("Invalid JSON for protocol headers, not a map")
+			}
+			for k, v := range h {
+				// fmt.Printf("HEADER: %v: %v\n", k, v)
+				// fmt.Printf("%v", reflect.TypeOf(v))
+				harray, ok := v.([]interface{})
+				if !ok {
+					return fmt.Errorf("Invalid JSON for protocol headers, not an array of strings for header value")
+				}
+				for _, vv := range harray {
+					rw.Header().Add(k, vv.(string)) // on top of any specified on the route
+				}
+			}
+		}
+	}
+	// after other header setting, top level content_type takes precedence and is
+	// absolute (if set). it is expected that if users want to set multiple
+	// values they put it in the string, e.g. `"content-type:"application/json; charset=utf-8"`
+	// TODO this value should not exist since it's redundant in proto headers?
+	if jout.ContentType != "" {
+		rw.Header().Set("Content-Type", jout.ContentType)
+	}
+
+	// we must set all headers before writing the status, see http.ResponseWriter contract
+	if p != nil && p["status_code"] != nil {
+		sc := p["status_code"].(float64)
+		rw.WriteHeader(int(sc))
+	}
+
+	if ci.IsCloudEvent() {
+		// then it's already in the right format so just return it as is
+		d, err := json.Marshal(jout)
+		if err != nil {
+			return fmt.Errorf("Error marshalling CloudEvent response to json. %v\n", err)
+		}
+		_, err = rw.Write(d)
```

Comment:
`json.NewEncoder(rw).Encode(jout)` saves some bytes, afaict from http code there's no difference as both will write into the response writers `bufio.Writer` and there's no framing risk (afaict)

Reply:
Done.